### PR TITLE
adding confirm_req to_string()

### PR DESCRIPTION
**Pull request by not-Ian**
re: #3960

### CODE REVIEW

File: nano/node/messages.cpp [hunk 1]
```diff
@@ -940,6 +940,21 @@ std::size_t nano::confirm_req::size (nano::block_type type_a, std::size_t count)
 	return result;
 }
 
+std::string nano::confirm_req::to_string () const
+{
+	std::stringstream stream;
+
+	stream << header.to_string ();
+
+	for (auto roots_hash = roots_hashes.begin (), end = roots_hashes.end (); roots_hash != end; ++roots_hash)
```
Comment (dsiganos) [inline nano/node/messages.cpp]: This line is too complicated. Please use range loop to simplify this code.
https://en.cppreference.com/w/cpp/language/range-for

File: nano/core_test/message.cpp [hunk 2]
```diff
@@ -243,6 +243,29 @@ TEST (message, confirm_req_hash_batch_serialization)
 	ASSERT_EQ (header.count_get (), req.roots_hashes.size ());
 }
 
+TEST (message, confirm_req_to_string)
```
Comment (dsiganos) [inline nano/core_test/message.cpp]: I asked for unit tests and thank you for doing them but let's make them minimal.
It is not worth having and maintaining large amount of code testing to_string functions.
Please re-write this to be minimal.

File: nano/core_test/message.cpp [hunk 3]
```diff
@@ -243,6 +243,29 @@ TEST (message, confirm_req_hash_batch_serialization)
...
+{
+	nano::send_block block = nano::send_block ();
+	std::shared_ptr block_ptr = std::make_shared<nano::send_block> (block);
+
+	nano::work_thresholds work_threshold = nano::work_thresholds (0, 0, 0);
```
Comment (dsiganos) [inline nano/core_test/message.cpp]: There is a much more concise way to do this, which we used in the test case:
'message.keepalive_to_string'. Please use that approach.

File: nano/core_test/message.cpp [hunk 4]
```diff
@@ -243,6 +243,29 @@ TEST (message, confirm_req_hash_batch_serialization)
...
+	nano::network_constants network_constants = nano::network_constants (work_threshold, nano::networks::nano_dev_network);
+	nano::confirm_req confirm_req = nano::confirm_req (network_constants, block_ptr);
+
+	std::string expected_string = "NetID: 5241(dev), VerMaxUsingMin: 19/19/18, MsgType: 4(confirm_req), Extensions: 0200";
+
+	for (auto roots_hash = 0, roots_hash_end = 3; roots_hash != roots_hash_end; ++roots_hash)
```
Comment (dsiganos) [inline nano/core_test/message.cpp]: There is no need for a for loop here. This is test code, expand the loop and do it manually.

File: nano/node/messages.cpp [hunk 5]
```diff
@@ -940,6 +940,21 @@ std::size_t nano::confirm_req::size (nano::block_type type_a, std::size_t count)
 	return result;
 }
 
+std::string nano::confirm_req::to_string () const
```
Comment (dsiganos) [inline nano/node/messages.cpp:943]: confirm_req has 2 modes of operation:
* list of hash pair
* block to confirm
You are only handling the list of pairs mode.
You need to handle both modes.

The block_type in the header determines which mode it is in.

File: nano/node/messages.cpp (hunk 1 above)
Comment (not-Ian) [reply nano/node/messages.cpp]: this should be resolved after the recent push

**Review by dsiganos — APPROVED** [review]

**Pull request merged**